### PR TITLE
PypeIt Class

### PR DESCRIPTION
**Pull request by profxj**
Significant refactor to eliminate ```arms.py``` and the previous ```pypeit.py```
and replace with one class to rule them all.

Lots of polishing to do before this PR should be merged (starting now).
But I should like a code review by @jhennawi and @kbwestfall at this stage.

And, everyone is encouraged (required) to branch of this branch for top-level development.

### CODE REVIEW

File: pypeit/core/fsort.py [hunk 1]
```diff
@@ -1107,14 +1107,17 @@ def make_dirs(spectrograph, caldir, scidir, qadir, overwrite=False):
...
             Flag to overwrite any existing files/directories.
     """
 
     # First, get the current working directory
-    currDIR = os.getcwd()
+    if redux_path is None:
+        redux_path = os.getcwd()
     msgs.info("Creating Science directory")
-    newdir = "{0:s}/{1:s}".format(currDIR, scidir)
+    newdir = "{0:s}/{1:s}".format(redux_path, scidir)
```
Comment (kbwestfall) [inline pypeit/core/fsort.py]: Use os.path.join.

File: pypeit/par/pypeitpar.py [hunk 2]
```diff
@@ -944,6 +944,10 @@ def __init__(self, spectrograph=None, pipeline=None, detnum=None, sortroot=None,
         descr['qadir'] = 'Directory relative to calling directory to write quality ' \
                          'assessment files.'
 
+        defaults['redux_path'] = ''
```
Comment (kbwestfall) [inline pypeit/par/pypeitpar.py]: Should we change this to cwd or something?

File: pypeit/pypeit.py [hunk 3]
```diff
@@ -4,157 +4,505 @@
...
-# TODO: (KBW) Can archeck code be put in pypeit/__init__.py ?
-from pypeit import check_requirements  # THIS IMPORT DOES THE CHECKING.  KEEP IT
 from pypeit.core import fsort
 from pypeit.core import qa
 from pypeit import arms
+from pypeit.par.util import make_pypeit_file, parse_pypeit_file
 
 from pypeit import pypeitsetup
+from pypeit.spectrographs import keck_lris
+from pypeit.scripts import run_pypeit
```
Comment (kbwestfall) [inline pypeit/pypeit.py]: Is this circular?

File: pypeit/pypeit.py [hunk 4]
```diff
@@ -4,157 +4,505 @@
...
+        sci_dict = OrderedDict()  # This needs to be ordered
+        sci_dict['meta'] = {}
+        sci_dict['meta']['vel_corr'] = 0.
+        #
+        scidx = np.where((self.fitstbl['sci_ID'] == sci_ID) & self.fitstbl['science'])[0][0]
+        msgs.info("Reducing file {0:s}, target {1:s}".format(self.fitstbl['filename'][scidx],
+                                                             self.fitstbl['target'][scidx]))
+
+        # Loop on Detectors
+        for kk in range(self.spectrograph.ndet):
```
Comment (jhennawi) [inline pypeit/pypeit.py]: How will you deal with a situation like LRIS red where the number of detectors is changing? I guess the spectrograph class needs to know something about the file then. The fundamental problem is we cannot know ndet until we've seen a file.

File: pypeit/core/fsort.py (hunk 1 above)
Comment (profxj) [reply pypeit/core/fsort.py]: done

File: pypeit/par/pypeitpar.py (hunk 2 above)
Comment (profxj) [reply pypeit/par/pypeitpar.py]: yup, done

File: pypeit/pypeit.py (hunk 3 above)
Comment (profxj) [reply pypeit/pypeit.py]: removed

File: pypeit/pypeit.py [hunk 5]
```diff
@@ -4,157 +4,706 @@
...
+        self.spectrograph = spectrograph
+        self.verbosity = verbosity
+        self.overwrite = overwrite
+        if setups_path is None:
+            self.setups_path = os.getcwd()
+        else:
+            self.setups_path = setups_path
+
+        # Internals
+        self.pypeit_file = ''
```
Comment (kbwestfall) [inline pypeit/pypeit.py]: Or `None` ?

**Comment by profxj** [comment]
Appear to be passing it all that it should.  Merging.

**Pull request merged**